content/post/new-python-api.md

```diff
@@ -0,0 +1,263 @@
+
+One comment was that the API was verbose. It required the user to create lots of classes, many of which had one method for doing something and another method for returning a name. In other words, many of these classes were really just functions with names. People thought that it was silly to have to define an entire class just to get a function. And they were right.
+
+Another comment was that there were places where a developer would have to write the same code every time they implemented something, even though for the most part the code was the same. For example, in decoders the user had to provide methods that:
+
+1. return the number of bytes that would represent the header for a message
+2. take an array of bytes and return the actual size of the message
+3. take the message bytes and return the object represented by the message
+
+Here’s what a decoder used to look like:
+
+```pony
+class Decoder(object):
+    def header_length(self):
+        return 4
+
+    def payload_length(self, bs):
+        return struct.unpack(">I", bs)[0]
+
+    def decode(self, bs):
+        return bs.decode("utf-8")
+```
+
+The only part of the decoder that really needs to have any logic is the part that takes bytes and returns a message; the first item can be described by an integer and the second one can be described by a string that is passed as an argument to `struct.unpack(...)` to tell it how to turn the incoming bytes into a number. With these changes, a decoder now looks like this:
+```python
+@wallaroo.decoder(header_length=4, length_fmt=">I")
+def decoder(bs):
+    return bs.decode("utf-8")
+```
+
+## Why We Created a New API
+
+Before we had even published the blog post about the original Python API, we had discussed ideas for things that we could do to improve the API. One of the ideas that we had considered was using decorators to cut down on the number of classes and thus reduce the amount of code that needed to be written. When we asked for feedback, several people suggested using decorators to improve the API, so we felt that our earlier idea had been validated. We went ahead and designed a new decorator-based API.
+
+## The New Python API
+
+### A Motivating Example
+
+We will start with the canonical streaming data processing application, Word Count. A stream of input text is analyzed and the total number of times each word has been seen is reported. [The example](https://github.com/WallarooLabs/wallaroo/blob/release/examples/python/word_count/word_count.py) in it's entirety is in our [GitHub repository](https://github.com/WallarooLabs/wallaroo/tree/release).
+
+We will make the following assumptions:
+* Incoming messages will come from a TCP connection and be sent to
```

Comment:
These bullets didn't render correctly in the preview.

Reply:
Fixing list.